File: android/src/main/kotlin/com/squareup/square_mobile_payments_sdk/modules/PaymentModule.kt
package com.squareup.square_mobile_payments_sdk.modules

import io.flutter.plugin.common.MethodChannel

import com.squareup.sdk.mobilepayments.MobilePaymentsSdk
import com.squareup.sdk.mobilepayments.core.Result as SdkResult
import com.squareup.sdk.mobilepayments.payment.Payment

import com.squareup.square_mobile_payments_sdk.mappers.PaymentMapper
import com.squareup.square_mobile_payments_sdk.extensions.toOfflineMap
import com.squareup.square_mobile_payments_sdk.extensions.toOnlineMap
import com.squareup.square_mobile_payments_sdk.extensions.toMoneyMap
import com.squareup.square_mobile_payments_sdk.extensions.toOfflineMap
import com.squareup.square_mobile_payments_sdk.extensions.toPaymentErrorCodeName
import com.squareup.square_mobile_payments_sdk.extensions.toErrorDetailsMap

class PaymentModule {
    companion object {
        private val paymentManager = MobilePaymentsSdk.paymentManager()

        @JvmStatic
        fun startPayment(
            result: MethodChannel.Result,
            paymentParameters: HashMap<String, Any>?,
            promptParameters: HashMap<String, Any>?
        ) {
            if (paymentParameters == null || promptParameters == null) {
                result.error(
                    "missingParameters",
                    "paymentParameters or promptParameters must not be null",
                    null
                )
                return
            }
            val nativePaymentParameters = PaymentMapper.getPaymentParameters(paymentParameters)
            val nativePromptParameters = PaymentMapper.getPromptParameters(promptParameters)

            paymentManager.startPaymentActivity(nativePaymentParameters, nativePromptParameters) { sdkResult ->
                when (sdkResult) {
                    is SdkResult.Success -> {
                        when (val payment = sdkResult.value) {
                            is Payment.OnlinePayment -> {
                                val mappedPayment = payment.toOnlineMap()
                                result.success(mappedPayment)
                            }
                            is Payment.OfflinePayment -> {
                                val mappedPayment = payment.toOfflineMap()
                                result.success(mappedPayment)
                            }
                        }
                    }
                    is SdkResult.Failure -> {
                        result.error(
                            sdkResult.errorCode.toPaymentErrorCodeName(),
                            sdkResult.errorMessage,
                            sdkResult.details.map { d -> d.toErrorDetailsMap() }
                        )
                    }
                    else -> {
                        result.error("unknown", null, null)
                    }
                }
            }
        }

        @JvmStatic
        fun getTotalStoredPaymentAmount(result: MethodChannel.Result) {
            val offlinePaymentQueue = paymentManager.getOfflinePaymentQueue()
            val sdkResult = offlinePaymentQueue.getTotalStoredPaymentAmount()
            when (sdkResult) {
                is SdkResult.Success -> {
                    result.success(sdkResult.value.toMoneyMap())
                }
                is SdkResult.Failure -> {
                    result.error(
                        sdkResult.errorCode.toPaymentErrorCodeName(),
                        sdkResult.errorMessage,
                        sdkResult.details.map { d -> d.toErrorDetailsMap() }
                    )
                }
            }
        }

        @JvmStatic
        fun getPayments(result: MethodChannel.Result) {
            val offlinePaymentQueue = paymentManager.getOfflinePaymentQueue()
            offlinePaymentQueue.getPayments { sdkResult ->
                when (sdkResult) {
                    is SdkResult.Success -> {
                        val paymentList = ArrayList<Map<String, Any?>>()
                        sdkResult.value.forEach { payment ->
                            paymentList.add(payment.toOfflineMap().mapKeys { it.key.toString() })
                        }
                        result.success(paymentList)
                    }
                    is SdkResult.Failure -> {
                        result.error(
                            sdkResult.errorCode.toPaymentErrorCodeName(),
                            sdkResult.errorMessage,
                            sdkResult.details.map { d -> d.toErrorDetailsMap() }
                        )
                    }
                }
            }
        }
    }
}

File: ios/Classes/Modules/PaymentModule.swift
import UIKit
import Flutter
import SquareMobilePaymentsSDK

public class PaymentModule: PaymentManagerDelegate {
    private static let paymentManager = MobilePaymentsSDK.shared.paymentManager
    private static let paymentDelegate = PaymentModule()
    private var delegateResult: FlutterResult?

    public static func startPayment(
        result: @escaping FlutterResult,
        paymentParameters: [String: Any],
        promptParameters: [String: Any]
        ) {
        let nativePaymentParameters = PaymentMapper.getPaymentParameters(paymentParameters:paymentParameters)
        let nativePromptParameters = PaymentMapper.getPromptParameters(promptParameters:promptParameters)

        guard let topController = UIApplication.shared.keyWindow?.rootViewController else {
            result(FlutterError(
                code: "notRootViewController",
                message: "No root view controller in window iOS app",
                details: nil))
            return
        }

        if (paymentDelegate.delegateResult != nil) {
            result(FlutterError(
                code: PaymentError.paymentAlreadyInProgress.getName(),
                message: "A payment is already in progress",
                details: nil))
            return
        }

        paymentDelegate.delegateResult = result
        paymentManager.startPayment(
            nativePaymentParameters,
            promptParameters: nativePromptParameters,
            from: topController,
            delegate: paymentDelegate
        )
    }

    public static func getPayments(result: @escaping FlutterResult) {
        let offlinePaymentQueue = paymentManager.offlinePaymentQueue
        offlinePaymentQueue.getPayments { payments, error in
            if let error = error {
                result(FlutterError(code: "GET_OFFLINE_PAYMENTS_FAILED", message: error.localizedDescription, details: nil))
            } else {
                let paymentsArray = payments.map { $0.toMap() }
                result(paymentsArray)
            }
        }
    }

    public static func getTotalStoredPaymentAmount(result: @escaping FlutterResult) {
        let offlinePaymentQueue = paymentManager.offlinePaymentQueue
        offlinePaymentQueue.getTotalStoredPaymentsAmount { moneyAmount, error in
            if let error = error {
                result(FlutterError(code: "GET_TOTAL_STORED_PAYMENTS_FAILED", message: error.localizedDescription, details: nil))
            } else if let moneyAmount = moneyAmount {
                result(moneyAmount.toMap())
            } else {
                //NEVER: if money is nil there was an error, so the error if will occur
                result(NSNull())
            }
        }
    }

    public func paymentManager(_ paymentManager: PaymentManager, didFinish payment: Payment) {
        if let onlinePayment = payment as? OnlinePayment {
            delegateResult?(onlinePayment.toMap())
        } else if let offlinePayment = payment as? OfflinePayment {
            delegateResult?(nil)
        } else {
            delegateResult?(nil)
        }
        delegateResult = nil
    }

    public func paymentManager(_ paymentManager: PaymentManager, didFail payment: Payment, withError error: Error) {
        let e = error as NSError
        if let paymentError = PaymentError(rawValue: e.code) {
            delegateResult?(FlutterError(
                code: paymentError.getName(),
                message: e.localizedDescription,
                details: e.localizedFailureReason
            ))
        } else {
            delegateResult?(FlutterError(
                code: "unknown",
                message: nil,
                details: nil
            ))
        }
        delegateResult = nil
    }

    public func paymentManager(_ paymentManager: PaymentManager, didCancel payment: Payment) {
        delegateResult?(FlutterError(
            code: "canceled",
            message: "the payment was cancelled",
            details: nil
        ))
        delegateResult = nil
    }

    // Optional
    public func paymentManager(_ paymentManager: PaymentManager, didStart payment: Payment) {
        print("Payment started.")
    }

    public func paymentManager(_ paymentManager: PaymentManager, willFinish payment: Payment) {
        print("Payment is about to finish.")
    }

    public func paymentManager(_ paymentManager: PaymentManager, willCancel payment: Payment) {
        print("Payment cancellation is in progress.")
    }
}